### PR TITLE
Custom Menubar Implementation

### PR DESCRIPTION
**Pull request by sbatten**
Adding custom menubar on windows with themable menus including context menus.

### CODE REVIEW

File: src/vs/platform/windows/common/windowsIpc.ts [hunk 1]
```diff
@@ -285,6 +300,14 @@ export class WindowsChannelClient implements IWindowsService {
 		return this.channel.call('unmaximizeWindow', windowId);
 	}
 
+	minimizeWindow(windowId: number): TPromise<void> {
+		return this.channel.call('minimizeWindow', windowId);
+	}
+
+	addMaximizeListener(windowId: number, listener: (maximized: boolean) => void): TPromise<void> {
+		return TPromise.as(null);
```
Comment (bpasero) [inline src/vs/platform/windows/common/windowsIpc.ts]: @sbatten what about this one, does it need a proper implementation?

File: src/vs/workbench/browser/parts/sidebar/sidebarPart.ts [hunk 2]
```diff
@@ -91,6 +91,7 @@ export class SidebarPart extends CompositePart<Viewlet> {
 		const container = $(this.getContainer());
 
 		container.style('background-color', this.getColor(SIDE_BAR_BACKGROUND));
+
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/sidebar/sidebarPart.ts]: @sbatten unrelated change?

File: src/vs/workbench/browser/parts/editor/editor.contribution.ts [hunk 3]
```diff
@@ -371,6 +371,7 @@ registry.registerWorkbenchAction(new SyncActionDescriptor(EditorLayoutTwoByTwoGr
 registry.registerWorkbenchAction(new SyncActionDescriptor(EditorLayoutTwoColumnsRightAction, EditorLayoutTwoColumnsRightAction.ID, EditorLayoutTwoColumnsRightAction.LABEL), 'View: Two Columns Right Editor Layout', category);
 registry.registerWorkbenchAction(new SyncActionDescriptor(EditorLayoutTwoColumnsBottomAction, EditorLayoutTwoColumnsBottomAction.ID, EditorLayoutTwoColumnsBottomAction.LABEL), 'View: Two Columns Bottom Editor Layout', category);
 registry.registerWorkbenchAction(new SyncActionDescriptor(EditorLayoutCenteredAction, EditorLayoutCenteredAction.ID, EditorLayoutCenteredAction.LABEL), 'View: Centered Editor Layout', category);
+registry.registerWorkbenchAction(new SyncActionDescriptor(EditorToggleAutoSaveAction, EditorToggleAutoSaveAction.ID, EditorToggleAutoSaveAction.LABEL), 'Toggle Auto Save', category);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editor.contribution.ts]: @sbatten I suggest to move this into `files.contribution.ts` and the related action into `fileActions.ts` because that seems like a better fit for a auto save related toggle action. I also think the category should then be "File" and not "View".

File: src/vs/platform/actions/common/actions.ts [hunk 4]
```diff
@@ -61,6 +61,17 @@ export class MenuId {
 	static readonly ViewItemContext = new MenuId();
 	static readonly TouchBarContext = new MenuId();
 	static readonly SearchContext = new MenuId();
+	static readonly MenubarFileMenu = new MenuId();
```
Comment (bpasero) [inline src/vs/platform/actions/common/actions.ts]: @sbatten what about the other sub-menus that we need to support, like "Editor Layout" (which was just added) and the ones under the "Go" menu? I think we should make sure that they all are covered as well. I would probably also add a TODO to keep in mind that a better solution would be to implement them via some real sub-menu support in the menus.

File: src/vs/code/electron-main/window.ts [hunk 5]
```diff
@@ -191,6 +195,9 @@ export class CodeWindow implements ICodeWindow {
 			this._win.setSheetOffset(22); // offset dialogs by the height of the custom title bar if we have any
 		}
 
+		this._win.on('maximize', (e) => app.emit('browser-window-maximize', e, this._win));
```
Comment (bpasero) [inline src/vs/code/electron-main/window.ts]: @sbatten move listeners into `registerListeners`

File: src/vs/code/electron-main/window.ts [hunk 6]
```diff
@@ -167,9 +167,10 @@ export class CodeWindow implements ICodeWindow {
 		}
 
 		let useCustomTitleStyle = false;
-		if (isMacintosh && (!windowConfig || !windowConfig.titleBarStyle || windowConfig.titleBarStyle === 'custom')) {
+		if ((!isMacintosh && windowConfig && windowConfig.titleBarStyle === 'custom') ||
```
Comment (bpasero) [inline src/vs/code/electron-main/window.ts]: @sbatten very hard to read this check, maybe extract some booleans from the windowConfig title style to make this better?

File: src/vs/base/browser/ui/actionbar/actionbar.ts [hunk 7]
```diff
@@ -273,7 +277,9 @@ export class ActionItem extends BaseActionItem {
 
 	public _updateLabel(): void {
 		if (this.options.label) {
-			this.$e.text(this.getAction().label);
+			let label = this.getAction().label;
```
Comment (bpasero) [inline src/vs/base/browser/ui/actionbar/actionbar.ts]: @sbatten shouldn't the concept of mnemonics only be used when the action shows up in a menu? It seems to me your changes make this happen anywhere the action is showing up. Maybe this should be an option for the action item to pass through to enable this.

File: src/vs/workbench/electron-browser/workbench.ts [hunk 8]
```diff
@@ -343,7 +351,14 @@ export class Workbench extends Disposable implements IPartService {
 		serviceCollection.set(IListService, this.instantiationService.createInstance(ListService));
 
 		// Context Menu
-		serviceCollection.set(IContextMenuService, new SyncDescriptor(ContextMenuService));
+		const hasCustomTitle = this.getCustomTitleBarStyle() === 'custom';
+
+		// Use themable context menus when custom titlebar is enabled to match custom menubar
+		if (isWindows && hasCustomTitle) {
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten why not linux here? It seems to me your custom menu can work on Linux too?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 9]
```diff
@@ -343,7 +351,14 @@ export class Workbench extends Disposable implements IPartService {
...
+			serviceCollection.set(IContextMenuService, new SyncDescriptor(HTMLContextMenuService, undefined, undefined, this.notificationService, this.contextViewService));
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten the 2x `undefined` seems wrong

File: src/vs/workbench/electron-browser/workbench.ts [hunk 10]
```diff
@@ -493,6 +511,8 @@ export class Workbench extends Disposable implements IPartService {
 			}
 		}
 
+		this.menubarPart.updateStyles();
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten I would let the menu bar part listen on the fullscreen change itself to `updateStyles`.

File: src/vs/workbench/browser/parts/menubar/media/menubarpart.css [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	padding: 0px 5px;
+	position: relative;
+	z-index: 10;
+	cursor: pointer;
+	-webkit-app-region: no-drag;
+}
+
+.monaco-workbench > .part.menubar > .menubar-menu-button.open,
+.monaco-workbench > .part.menubar > .menubar-menu-button:hover {
+	background-color: rgba(255, 255, 255, 0.1);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/menubar/media/menubarpart.css]: @sbatten any color defined in CSS needs to be externalized to a themable color in [`theme.ts`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/common/theme.ts#L1) so that theme authors can tweak them.

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 12]
```diff
@@ -5,6 +5,7 @@
 
 'use strict';
 
+import * as path from 'path';
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts]: @sbatten layer breaker, you cannot have a node.js dependency in `browser` code. there is a `vs/base/path` library you can use instead.

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 13]
```diff
@@ -228,6 +230,16 @@ export class TitlebarPart extends Part implements ITitleService {
 	public createContentArea(parent: HTMLElement): HTMLElement {
 		this.titleContainer = $(parent);
 
+		if (!isMacintosh) {
+			$(this.titleContainer).img({
+				class: 'window-appicon',
+				src: path.join(this.environmentService.appRoot, 'resources/win32/code.ico')
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts]: @sbatten does this work on Linux too? I think this is a Windows only format

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 14]
```diff
@@ -261,16 +307,47 @@ export class TitlebarPart extends Part implements ITitleService {
...
 	protected updateStyles(): void {
 		super.updateStyles();
 
 		// Part container
 		if (this.titleContainer) {
+			const bgColor = this.getColor(this.isInactive ? TITLE_BAR_INACTIVE_BACKGROUND : TITLE_BAR_ACTIVE_BACKGROUND);
 			this.titleContainer.style('color', this.getColor(this.isInactive ? TITLE_BAR_INACTIVE_FOREGROUND : TITLE_BAR_ACTIVE_FOREGROUND));
-			this.titleContainer.style('background-color', this.getColor(this.isInactive ? TITLE_BAR_INACTIVE_BACKGROUND : TITLE_BAR_ACTIVE_BACKGROUND));
+			this.titleContainer.style('background-color', bgColor);
+			this.titleContainer.getHTMLElement().classList.toggle('light', Color.fromHex(bgColor).isLighter());
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts]: @sbatten there is a `toggleClass` method on the builder directly

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 15]
```diff
@@ -250,6 +262,40 @@ export class TitlebarPart extends Part implements ITitleService {
 			}
 		});
 
+		if (!isMacintosh) {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts]: @sbatten there seems to be quite a bit of duplication between this code and `onDidChangeMaximized`. I suggest to move everything into `onDidChangeMaximized` and then call it from here with the initial value to prevent that.

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 16]
```diff
@@ -8,24 +8,133 @@
 	width: 100%;
 	font-size: 12px;
 	padding: 0 70px;
-	overflow: hidden;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @sbatten careful, does that mean the title value can now overflow if it is very long?

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 17]
```diff
@@ -8,24 +8,133 @@
...
 	text-overflow: ellipsis;
-	-webkit-app-region: drag;
 	zoom: 1; /* prevent zooming */
+}
+
+/***********************\
+| Windows and Linux CSS |
+\***********************/
+
+.monaco-workbench.windows > .part.titlebar, .monaco-workbench.linux > .part.titlebar {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @sbatten I think it is easier to read multiple CSS rules that are separated with "`,`" by putting each rule on a separate line

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 18]
```diff
@@ -8,24 +8,133 @@
...
+	order: 1;
+}
+
+.monaco-workbench > .part.titlebar.titlebar.light > .window-controls-container > .window-minimize {
+	background-image: url('chrome-minimize.svg');
+	order: 1;
+}
+
+.monaco-workbench > .part.titlebar > .window-controls-container > .window-icon:hover {
+	background-color: rgba(255, 255, 255, 0.1);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @sbatten again, all colors need to go into `theme.ts`

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 19]
```diff
@@ -8,24 +8,133 @@
...
+	-webkit-transition: background-color .2s;
+	transition: background-color .2s;
+	height: 95%;
+	width: 46px;
+	background-size: 10px;
+	background-position: center center;
+	background-repeat: no-repeat;
+}
+
+.window-unmaximize {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @sbatten I suggest to qualify this rule better

File: src/vs/workbench/browser/layout.ts [hunk 20]
```diff
@@ -102,6 +108,8 @@ export class WorkbenchLayout extends Disposable implements IVerticalSashLayoutPr
 		this.sashXTwo = new Sash(this.workbenchContainer, this);
 		this.sashY = new Sash(this.workbenchContainer, this, { orientation: Orientation.HORIZONTAL });
 
+		this.onMaximizeChange(false);
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @sbatten there is a listener missing to call this when the event happens?

File: src/vs/workbench/browser/layout.ts [hunk 21]
```diff
@@ -646,6 +665,7 @@ export class WorkbenchLayout extends Disposable implements IVerticalSashLayoutPr
 
 		// Propagate to Part Layouts
 		this.parts.titlebar.layout(new Dimension(this.workbenchSize.width, this.titlebarHeight));
+		this.parts.menubar.layout(new Dimension(undefined, this.menubarHeight));
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @sbatten I would put the width here too if it applies (possibly the width of the workbench?)

File: src/vs/workbench/browser/layout.ts [hunk 22]
```diff
@@ -655,8 +675,18 @@ export class WorkbenchLayout extends Disposable implements IVerticalSashLayoutPr
 		this.contextViewService.layout();
 	}
 
+	onMaximizeChange(maximized: boolean): void {
+		if (!isMacintosh) {
+			this.titlebarBaseHeight = maximized ? 23 : this.partLayoutInfo.titlebar.height;
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @sbatten where does the `23` come from? Should it be a constant?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 23]
```diff
@@ -1049,16 +1093,34 @@ export class Workbench extends Disposable implements IPartService {
...
 
 			// Update toasts
 			this.notificationsToasts.update(this.notificationsCenter.isVisible);
+
+			// Register commands
+			registerNotificationCommands(this.notificationsCenter, this.notificationsToasts);
 		}));
+	}
+
+	private _onTitleBarVisibilityChange: Emitter<void> = new Emitter<void>();
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten unrelated changes? Looks like these just moved some lines up?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 24]
```diff
@@ -1432,6 +1479,14 @@ export class Workbench extends Disposable implements IPartService {
 		this.workbenchLayout.layout();
 	}
 
+	setMenubarHidden(visibility: string, skipLayout: boolean): void {
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten maybe visibility could have a better type to reflect the values this can have (e.g. `visibility: 'visible' | 'default'`)

File: src/vs/workbench/electron-browser/workbench.ts [hunk 25]
```diff
@@ -925,6 +949,15 @@ export class Workbench extends Disposable implements IPartService {
 			this.notificationsCenter,
 			this.notificationsToasts
 		);
+
+		if (!isMacintosh && this.getCustomTitleBarStyle()) {
+			this.windowService.isMaximized().then((max) => {
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten I am not a big fan of the various calls to `windowService.isMaximized()` on startup because this runs async while the workbench is being created and can potentially lead into flicker. I would suggest to carry over a property into the `IWindowConfiguration` that has this information already in so that you do not have to call this on startup async. 

The layout should then just receive this as part of creation and do the right thing.

**Review by bpasero — COMMENTED** [review]
@sbatten some feedback provided. There seem to be compile errors:

image

Also, when I run this on mac, I get this error right on startup:

```
"TypeError: Cannot read property 'getContainer' of undefined
    at WorkbenchLayout.layout (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/browser/layout.js:484:55)
    at WorkbenchLayout.onMaximizeChange (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/browser/layout.js:588:18)
    at WorkbenchLayout [as constructor] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/browser/layout.js:67:19)
    at create (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/types.js:164:14)
```

File: src/vs/base/browser/ui/actionbar/actionbar.ts (hunk 7 above)
Comment (sbatten) [reply src/vs/base/browser/ui/actionbar/actionbar.ts]: here I followed the pattern I thought was present in the electron menu. If the action item label has &&, it will create a mnemonic. I only pass these in menu creation. I might be able to move this to ui/menu.ts if you prefer?

File: src/vs/workbench/electron-browser/workbench.ts (hunk 9 above)
Comment (sbatten) [reply src/vs/workbench/electron-browser/workbench.ts]: I added the telemetryservice, not sure if im supposed to pass something for the html container here. its supposed to be the context menu container but we don't have that yet, undefined/null seems to create it on the fly

File: src/vs/workbench/browser/layout.ts (hunk 20 above)
Comment (sbatten) [reply src/vs/workbench/browser/layout.ts]: its called from workbench.ts@960

File: src/vs/workbench/electron-browser/workbench.ts (hunk 25 above)
Comment (sbatten) [reply src/vs/workbench/electron-browser/workbench.ts]: I understand the complaint, but I don't quite understand the solution. The window service that holds the configuration still has to make an async call into the windowsService it holds

File: src/vs/workbench/electron-browser/workbench.ts (hunk 24 above)
Comment (sbatten) [reply src/vs/workbench/electron-browser/workbench.ts]: how do I then pass a string from the configuration to it? I union'd with string for the time being

File: src/vs/workbench/electron-browser/workbench.ts (hunk 25 above)
Comment (bpasero) [reply src/vs/workbench/electron-browser/workbench.ts]: @sbatten I suggest to transport this state over all the way from the place where the window loads the URL in the browser process: https://github.com/Microsoft/vscode/blob/master/src/vs/code/electron-main/window.ts#L521

It should be possible to know the initial state there and then you do not need to make an async call on workbench creation to query it. The config object is accessible for anyone via [IWindowService#getConfiguration()](https://github.com/Microsoft/vscode/blob/master/src/vs/platform/windows/electron-browser/windowService.ts#L36)

File: src/vs/base/browser/ui/actionbar/actionbar.ts (hunk 7 above)
Comment (sbatten) [reply src/vs/base/browser/ui/actionbar/actionbar.ts]: @bpasero in the interest of getting some insider testing, are you willing to put off changing this in the next check-in? I'd like to focus on items that need to be fixed for initial insider testing so I can then shift to the milestone goal (this included)

File: src/vs/base/browser/ui/menu/menu.css [hunk 26]
```diff
@@ -87,6 +87,10 @@
 	color: inherit;
 }
 
+.monaco-menu .monaco-action-bar.vertical .action-label.checked:after {
```
Comment (bpasero) [inline src/vs/base/browser/ui/menu/menu.css]: @sbatten is this used currently anywhere?

File: src/vs/code/electron-main/app.ts [hunk 27]
```diff
@@ -487,9 +493,6 @@ export class CodeApplication {
 			}
 		}
 
-		// Install Menu
-		appInstantiationService.createInstance(CodeMenu);
```
Comment (bpasero) [inline src/vs/code/electron-main/app.ts]: @sbatten is the functionality from `vs/code/electron-main/menus.ts` moved over to `menubar.ts`? then I think it would be safe to remove `menu.ts`?

However I see that some things are not working yet, like the Accessibility help window, so maybe that is still work in progress?

File: src/vs/workbench/browser/parts/editor/editorActions.ts [hunk 28]
```diff
@@ -1580,7 +1580,6 @@ export class EditorLayoutCenteredAction extends Action {
 	}
 
 	public run(): TPromise<any> {
-
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorActions.ts]: @sbatten unrelated change?

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 29]
```diff
@@ -24,8 +24,128 @@
 	flex: 0 1 auto;
 	overflow: hidden;
 	white-space: nowrap;
-	line-height: 22px;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @sbatten why? this impacts even macOS that does not use a custom menu

**Review by bpasero — COMMENTED** [review]
Some more code review feedback

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css (hunk 29 above)
Comment (sbatten) [reply src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css]: @bpasero the value is inherited from the parent and the parent is being updated to handle zoom

File: src/vs/workbench/browser/parts/editor/editorActions.ts (hunk 28 above)
Comment (sbatten) [reply src/vs/workbench/browser/parts/editor/editorActions.ts]: @bpasero reverted

File: src/vs/base/browser/ui/menu/menu.css (hunk 26 above)
Comment (sbatten) [reply src/vs/base/browser/ui/menu/menu.css]: @bpasero this adds the checkmark to items that have been checked in the menu, like auto save

**Review by bpasero — APPROVED** [review]
@sbatten I did another code review and pushed some polish to the branch. I think there is one thing to revisit before merging: the changes in `ActionBar` have some issues:
* we are no longer using `builder.text` but [`builder.innerHTML`](https://github.com/sbatten/vscode/blob/titlebarplus/src/vs/base/browser/ui/actionbar/actionbar.ts#L284) for the label. this increases the risk of someone injecting a HTML script into our DOM because we set it blindly as HTML. we need to continue to not trust `IAction.label` as before imho
* [`Actionbar#push()`](https://github.com/sbatten/vscode/blob/titlebarplus/src/vs/base/browser/ui/actionbar/actionbar.ts#L600) seems to do the mnemonic handling now. first of all, we should probably only do that if the action that is pushed has `isMenu: true` in the options. Second, there seems to be a [map of mnemonics](https://github.com/sbatten/vscode/blob/titlebarplus/src/vs/base/browser/ui/actionbar/actionbar.ts#L383) to `IAction` but the lifecycle is not clear to me. Actions can come and go in the action bar but this map seems to just grow and never cleanup when an action is removed?

**Comment by equinusocio** [comment]
@bpasero Hi, there is already a documentation for this api?
